zproject/backends.py

```diff
@@ -867,6 +888,62 @@ def sync_custom_profile_fields_from_ldap(
         except SyncUserException as e:
             raise ZulipLDAPException(str(e)) from e
 
+    def sync_groups_from_ldap(self, user_profile: UserProfile, ldap_user: _LDAPUser) -> None:
+        if not settings.LDAP_SYNC_GROUPS:
+            return
+
+        if user_profile.realm_id != settings.LDAP_GROUP_SYNC_REALM_ID:
+            return
+
+        try:
+            ldap_logger.debug("found groups for user: %s", str(user_profile))
+            ldap_group_name_set = set(ldap_user.group_names).intersection(set(settings.LDAP_GROUP_SYNC_LIST))
+
+            member_group_name_set = set(
+                [
+                    membership.user_group.name
+                    for membership in UserGroupMembership.objects.filter(
+                        user_group__realm=user_profile.realm,
+                        user_group__name__in=set(settings.LDAP_GROUP_SYNC_LIST),
+                        user_profile=user_profile,
+                    )
+                ]
+            )
+
+            ldap_logger.debug("ldap groups: %s", repr(ldap_group_name_set))
+            ldap_logger.debug("zulip groups: %s", repr(member_group_name_set))
+
+            add_group_list = []
+
+            for group_to_add in ldap_group_name_set.difference(member_group_name_set):
+                ldap_logger.debug("add %s to %s", user_profile.full_name, str(group_to_add))
+
+                group = UserGroup.objects.get(name=group_to_add, realm=user_profile.realm)
+
+                add_group_list.append(UserGroupMembership(
+                    user_group_id=group.id,
+                    user_profile_id=user_profile.id,
+                ))
+
+            UserGroupMembership.objects.bulk_create(add_group_list)
+
+            del_groupid_list = []
+
+            for group_to_remove in member_group_name_set.difference(ldap_group_name_set):
+                ldap_logger.debug("remove %s from %s", user_profile.full_name, group_to_remove)
+
+                group = UserGroup.objects.get(name=group_to_remove, realm=user_profile.realm)
```

Comment:
Same comment as above applies to do a query to fetch all the groups in one query rather in a loop with `.get()`s

Reply:
see above, re-written